### PR TITLE
Two factor autentication support

### PR DESCRIPTION
**Pull request by fredj**
* add 'should change password' in admin interface

### CODE REVIEW

File: commons/c2cgeoportal_commons/alembic/static/0c640a58a09a_add_opt_key_column.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+down_revision = 'ae5e88f35669'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    staticschema = config['schema_static']
+
+    op.add_column('user', Column('otp_key', String), schema=staticschema)
+    op.add_column('user', Column('temp_otp_key', String), schema=staticschema)
```
Comment (sbrunner) [inline commons/c2cgeoportal_commons/alembic/static/0c640a58a09a_add_opt_key_column.py]: Should be a tech_data column of type hstore...
Do you get an issue with that?

File: geoportal/c2cgeoportal_geoportal/views/entry.py [hunk 2]
```diff
@@ -1110,10 +1115,15 @@ def login(self):
...
         if username is not None:
-            user = models.DBSession.query(static.User).filter(static.User.username == username).one_or_none()
-            if user:
-                user.update_last_login()
-            headers = remember(self.request, username)
+            user = models.DBSession.query(static.User).filter(static.User.username == username).one()
+            user.update_last_login()
+            if self.two_factor_auth:
+                if not user.validate_2fa_totp(self.request.POST.get("otp")):
+                    raise HTTPUnauthorized("See server logs for details")
```
Comment (arnaud-morvan) [inline geoportal/c2cgeoportal_geoportal/views/entry.py]: Maybe user.update_last_login() after user.validate_2fa_totp ?

Comment (sbrunner) [reply geoportal/c2cgeoportal_geoportal/views/entry.py]: +1 :-)

**Comment by sbrunner** [comment]
Thanks @arnaud-morvan, updated :-)

**Comment by arnaud-morvan** [comment]
No chapter in doc ?

**Comment by sbrunner** [comment]
Yes => https://github.com/camptocamp/c2cgeoportal/pull/5198/files#diff-a923aa53e14368b38bf1b0acda7752fdR29 :-)

File: doc/integrator/authentication.rst [hunk 3]
```diff
@@ -26,3 +26,19 @@ For logging in, ``c2cgeoportal`` validates the user credentials
 table. If a c2cgeoportal application should work with another user information
 source, like LDAP, a custom *client validation* mechanism can be set up.
 Our knowledge base has an example of how this can be achieved.
+
+Tow factor autentication
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+GeoMapFish support TOTP (Time-Based One-Time Password Algorithm) tow factor autentication
```
Comment (arnaud-morvan) [inline doc/integrator/authentication.rst]: two factors authentication

File: doc/integrator/authentication.rst [hunk 4]
```diff
@@ -26,3 +26,19 @@ For logging in, ``c2cgeoportal`` validates the user credentials
...
+(`RFC 6238 <https://tools.ietf.org/html/rfc6238>`_).
+To enable the tow factor autentication you should set the following settings:
```
Comment (arnaud-morvan) [inline doc/integrator/authentication.rst]: two factors authentication

File: doc/integrator/authentication.rst [hunk 5]
```diff
@@ -26,3 +26,19 @@ For logging in, ``c2cgeoportal`` validates the user credentials
...
+
+.. code:: yaml
+
+   vars:
+     authentication:
+       two_factor: true
+
+If a user lost his secound autentication factor he can't ask for a new one, to reset it the administrator
+should uncheck the 'The user changed his password' field on the user in the admin interface.
```
Comment (arnaud-morvan) [inline doc/integrator/authentication.rst]: second authentication

**Comment by sbrunner** [comment]
Thanks @arnaud-morvan fixed :-)

**Pull request merged**